radix-engine/src/kernel/kernel.rs

```diff
@@ -98,18 +155,30 @@ impl<'h, M: KernelCallbackObject, S: SubstateDatabase> BootLoader<'h, M, S> {
             v.borrow_mut();
         });
 
+        // Read kernel boot configuration
+        let kernel_boot: KernelBoot = self
+            .track
+            .read_boot_substate(
+                TRANSACTION_TRACKER.as_node_id(),
+                BOOT_LOADER_PARTITION,
+                &SubstateKey::Field(BOOT_LOADER_KERNEL_BOOT_FIELD_KEY),
+            )
+            .map(|v| scrypto_decode(v.as_slice()).unwrap())
+            .unwrap_or(KernelBoot::babylon());
+
         // Create System
         let mut callback = M::init(&mut self.track, executable, self.init.clone())?;
 
         // Create Kernel
         let mut kernel = {
-            let mut kernel = Kernel::new(&mut self.track, &mut self.id_allocator, &mut callback);
-
             // Check references
-            let engine_references = kernel
-                .check_references(executable.references())
+            let costing_enabled = kernel_boot.is_ref_check_costing_enabled();
+            let engine_references = self
+                .check_references(&mut callback, executable.references(), costing_enabled)
```

Comment:
I do like the idea of loading the kernel boot substate upfront, but I'm not sure about the value proposition of moving reference checks to bootloading. It couples a lot of logic, e.g. the direct reference recognition to bootloading. Also, we're directly manipulating the call frame of kernel afterwards anyway.